Refactor tkey-sign in the spirit of signify

mchack-work opened the pull request.

dehanj commented on cmd/tkey-sign/main.go:
```diff
@@ -0,0 +1,389 @@
+
+%[1]s -h/--help
+
+%[1]s -G/--getkey [-d/--port device] [-f/--force] [-s/--speed speed] [--uss] [--uss-file ussfile] [--verbose] -p/--public pubkey
```
Maybe `-p/--public pubkey` should be directly after `-G/--getkey`. The all optional can come afterwards, easier to see what is needed directly.

mchack-work replied:
Done. And sorted alphabetically.

dehanj commented on cmd/tkey-sign/main.go:
```diff
@@ -0,0 +1,389 @@
+	enterUss := pflag.Bool("uss", false,
+		"Enable typing of a phrase to be hashed as the User Supplied Secret. The USS is loaded onto the TKey along with the app itself. A different USS results in different public/private keys, meaning a different identity.")
+	ussFile := pflag.String("uss-file", "",
+		"Read `FILE` and hash its contents as the USS. Use '-' (dash) to read from stdin. The full contents are hashed unmodified (e.g. newlines are not stripped).")
```
We are not consistent with if the input/output file for an argument are capitalized or not.
Example: 
```
-m, --message string   Message file
-d, --port PATH        Set serial port device PATH. If this is not passed,
 ...
--uss-file FILE    Read FILE and hash its contents as the USS. Use '-'
                         (dash) to read from stdin. The full contents are hashed
                         unmodified (e.g. newlines are not stripped).
 ```
Not sure which one is most common to use. Some use <file> when it is a placeholder...

mchack-work replied:
Now consistent.

dehanj commented on cmd/tkey-sign/main.go:
```diff
@@ -0,0 +1,389 @@
+	switch cmd {
+	case cmdGetKey:
+		if *keyFile == "" {
+			le.Printf("Provide public key file with -p FILE")
```
Here we say "Provide public key file with -p FILE".
While in the examples we say: "-p/--public pubkey"
and in the usage description we say: "-p, --public string"

We should use the same on all three instead of FILE, pubkey and string.

mchack-work replied:
I harmonized man page, help text and the error messages.

dehanj commented on cmd/tkey-sign/main.go:
```diff
@@ -0,0 +1,389 @@
+		}
+
+		if *sigFile == "" {
+			*sigFile = *messageFile + ".sig"
```
Maybe we should check if that exists, and if not say "Signature file not found, specify with '-x <file>'" or something. Would be nicer than the the error on row 377 today:
"Error verifying: decodeFileInput: could not read Makefile.sig: open Makefile.sig: no such file or directory".
which absolutely tells precisely the issue. Maybe not super important, but would make it a bit nicer with an instruction in the right direction. 

One could use Stat instead of open to accomplish it if one wants to fail fast:
```
_, err := os.Stat(*sigFile)
if err != nil {
  if os.IsNotExist(err) {
    // File or directory does not exist
  } else {
    // Some other error. The file may or may not exist
  }
}
```

mchack-work replied:
I added a check checking the error from readSig() to see if it's ErrNotExist and give the useful hint.

dehanj commented on cmd/tkey-sign/main.go:
```diff
@@ -0,0 +1,389 @@
+		if err != nil {
+			le.Printf("signing failed: %v", err)
+			signer.Close()
+			os.Exit(1)
```
We could create our own exit function that does the closing, and then uses os.Exit(x). Would make it less error prone. 
Especially if we can run close even if it is not open. But might not be so important to fix right now.

mchack-work replied:
We can't yet close if it's not open without changes to tkeysign. There's an issue for that in tkeysign. Resolving.

dehanj commented:
I did some more testing, mainly I wanted to see if it is possible to verify our signatures using signify.
From start it is not possible right away, but doing a sha512sum on the message, and trying to verify the signature on that, should work, right? Since we are signing the sha512sum. 

I faced two issues;
1. Signify is adding a comment to each file:
Something in line with "untrusted comment: signify public key" depending on what the file includes, and it does not accept the file if it does not start with "untrusted comment:". I added this to the files generated by tkey-sign-cli, to test, rendering issue 2 below.

2. It uses a different base64 encoding.
```
signify -V -m Makefile -x Makefile.sig -p pubkey
signify: invalid base64 encoding in pubkey
```
Same issue for the "Makefile.sig". 

So it seems that `encoding/base64` uses: "Package base64 implements base64 encoding as specified by [RFC 4648](https://rfc-editor.org/rfc/rfc4648.html)."

and signify are using: [RFC1521](https://www.rfc-editor.org/rfc/rfc1521), or MIME Base64 Content-Transfer-Encoding.

Should we use the same? Or have a flag to use the same?

mchack-work commented:
I think the idea of being compatible with signify to the point where we can use signify to verify a signature made by tkey-sign is good so I have added that functionality. It required adding a comment line and metadata to the stored pubkey and signature.

Since we use a hash digest you must also run `sha512sum` or similar before calling `signify`. To make it easier I replaced our binary digest with a hex digest like the output of `sha512sum`.

There's now a script `signify-verify` that automatically calls `sha512sum` and then `signify -V`.

mchack-work commented:
CI currently fails with a cast problem but works locally. Probably a different Go compiler.

mchack-work commented:
CI fixed by using `copy()` for now.

dehanj commented on signify-verify:
```diff
@@ -0,0 +1,22 @@
+
+digestfile=$(mktemp /tmp/tkey-sign-digest.XXXXXX)
+sigfile="${1}".sig
+sha512sum $1 > ${digestfile}
```
sha512sum is linux only. Do we want it running macos? 
We could use uname, or simply have an input..? 

But works great if I call `shasum -a 512 .. `
